test_plans/barriers_fences.asciidoc

```diff
@@ -0,0 +1,74 @@
+:sectnums:
+:xrefstyle: short
+
+= Test plan for barriers and fences
+
+This is a test plan for the APIs described in https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:barriers-fences[SYCL 2020 sections 4.15.1. Barriers and fences]
+and https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#_group_barrier[4.17.3.3 group_barrier]
+
+== Testing scope
+
+=== Device coverage
+
+All of the tests described below are performed only on the default device that
+is selected on the CTS command line.
+
+== Tests
+
+=== SYCL2020 interface
+
+==== Testing `sycl::atomic_fence` function

```

Comment:
I was having some difficulty following all the details of the `atomic_fence` test plan, but here are some general comments:

* I think we should concentrate on testing fence operations between work-items in a single kernel.  I think we can wait before adding tests for fence operations between different kernels, different devices, or between the host and device.
* We should avoid using USM allocations in this test because USM is an optional feature, which may not be supported by all implementations.  The test can use `local_accessor` memory to test synchronization between work-items in the same work-group and a buffer `accessor` to test synchronization between work-items in different work-groups.
* We can still test different memory scopes.  `memory_scope::system` and `memory_scope::device` can still be used to synchronize memory operations between work-items in the same kernel.
* Remember that all memory scopes aside from `memory_scope::work_group` are optional.  The test needs to use `info::device::atomic_fence_scope_capabilities` to see which other memory scopes are supported.
* The memory order `memory_order::seq_cst` is also an optional feature.  The test needs to use `info::device::atomic_fence_order_capabilities` to see if it is supported.
* I believe it is only safe to use `atomic_fence` in conjunction with atomic operations like `atomic_ref`.  Perhaps @Pennycook can provide a code snippet to use.

I also have a question for @Pennycook.  I see the spec guarantees that `atomic_fence` supports all memory orders except `memory_order::seq_cst`, which is an optional feature.  However, `atomic_ref` is only guaranteed to support `memory_order::relaxed`, and all the others are optional features.  Is that what we intended?  It seems a little weird that they are different.

One other suggestion ... I believe the existing CTS had a test for the SYCL 1.2.1 function `group::mem_fence`.  The semantics of that function are not exactly the same as `atomic_fence`, but you might be able to get some ideas from looking at that old test.

Reply:
> I believe it is only safe to use atomic_fence in conjunction with atomic operations like atomic_ref. Perhaps @Pennycook can provide a code snippet to use.

Right.  C++ defines two ways that fences can synchronize with one another [here](http://eel.is/c++draft/atomics.fences), but both require an atomic operation to some memory object _M_.

I don't have an example code ready to go, but I think it should be possible to construct examples by following the wording in the C++ specification.  For example, to test a release fence synchronizing with an acquire fence, as described below:

> A release fence _A_ synchronizes with an acquire fence _B_ if there exist atomic operations _X_ and _Y_, both operating on some atomic object _M_, such that _A_ is sequenced before _X_, _X_ modifies _M_, _Y_ is sequenced before _B_, and _Y_ reads the value written by _X_ or a value written by any side effect in the hypothetical release sequence _X_ would head if it were a release operation.

...we could do something like:

```c++
// work-item 0
data[0] = 42; // Non-atomic write to data[0]
sycl::atomic_fence(sycl::memory_order::release, sycl::memory_scope::work_group); // Fence A
sycl::atomic_ref<int, sycl::memory_order::relaxed, sycl::memory_scope::work_group>(M) = 1; // Atomic X

// work-item 1
int val = sycl::atomic_ref<int, sycl::memory_order::relaxed, sycl::memory_scope::work_group>(M); // Atomic Y
sycl::atomic_fence(sycl::memory_order::acquire, sycl::memory_scope::work_group); // Fence B
x = data[0]; // Non-atomic read of data[0]
if (val == 1) {
  assert(x == 42);
}
```

Because _A_ synchronizes with _B_, the write to data[0] from work-item 0 must be visible to work-item 1 if atomic X executes before atomic Y.

...I'm not sure how to make this into a good test, however, because there's no guarantee in the code above that work-item 0 executes before work-item 1 does.

> I also have a question for @Pennycook. I see the spec guarantees that atomic_fence supports all memory orders except memory_order::seq_cst, which is an optional feature. However, atomic_ref is only guaranteed to support memory_order::relaxed, and all the others are optional features. Is that what we intended? It seems a little weird that they are different.

I agree it is a little weird, and I don't think I fully understand why they're different.

I remember that SYCL 1.2.1's `mem_fence` supported "read" and "write" fences, and that we decided these mapped to acquire and release fences in C++11 terminology.  A relaxed fence doesn't do anything, so any device supporting only relaxed fences effectively wouldn't support memory fences at all.  I think we decided this didn't make sense, because all devices support `group_barrier` and `group_barrier` implies an acquire-release fence at work-group scope; a device that didn't support acquire-release fences wouldn't be able to implement SYCL.

I don't know why this same logic doesn't extend to supporting acquire-release for atomic operations, especially if they can be emulated by generating an `atomic_fence` before or after a relaxed atomic.  But this is another situation where the SYCL behavior is the same as OpenCL, so I'm hoping @bashbaug knows :smile:.

Reply:
Yes, this is slightly messy, but the minimum capabilities were determined based on the definitions of the older OpenCL 1.x atomic functions (e.g. `atomic_add`) and fence functions (e.g. `read_mem_fence`).  There is a detailed discussion about this here, with a reference to a related internal Khronos issue as well:

https://github.com/KhronosGroup/OpenCL-Docs/issues/354

In short, since one of the design goals of OpenCL 3.0 was to provide an easy path from OpenCL 1.2 to 3.0, we chose to separate the "atomic memory" capabilities from the "atomic fence" capabilities. This is another case where SYCL could choose to do something different and require the same (stronger) memory orders for both atomic_ref and atomic_fence, though it would then be SYCL's responsibility to insert the proper fences when a strong memory order is used with an atomic_ref that is only supported by an atomic_fence.

Reply:
In the test plan, I described similar to what @Pennycook is suggesting. But since there's indeed no guarantee that work-item 0 executes before work-item 1 does, I would suggest instead of:
x = data[0]; // Non-atomic read of data[0]
if (val == 1) {
  assert(x == 42);
}
Use loop:
while (1 != (val = sycl::atomic_ref<int, sycl::memory_order::relaxed, sycl::memory_scope::work_group>(M)));
x = data[0]; // Non-atomic read of data[0]
assert(x == 42);

@gmlueck 
Removed the statement about different kernels and devices from the test plan,
Removed the using of USM,
Added checks using accessor and local_accessor,
Added a statement about the requirement to check device capabilities.